File: Mak_V3/Chapter18/RecordTest.pas
PROGRAM RecordTest;

TYPE
    String2  = ARRAY [0..1 ] OF char;
    String5  = ARRAY [0..4 ] OF char;
    String8  = ARRAY [0..7 ] OF char;
    String16 = ARRAY [0..15] OF char;

    AddressRec = RECORD
                     street : String16;
                     city   : String16;
                     state  : String2;
                     zip    : String5;
                 END;

    PersonRec =  RECORD
                     firstName : String16;
                     lastName  : String16;
                     age       : integer;
                     address   : AddressRec;
                     phones    : ARRAY [0..1] OF String8;
                 END;

VAR
    john : PersonRec;
    persons : ARRAY [0..4] OF PersonRec;

    age    : integer;
    name   : String16;
    street : String16;

PROCEDURE print(VAR pers : PersonRec);

    VAR
        age  : integer;
        addr : AddressRec;
        phs  : ARRAY [0..1] OF String8;

    BEGIN
        age := pers.age;

        addr.street := pers.address.street;
        addr.city   := pers.address.city;
        addr.state  := pers.address.state;
        addr.zip    := pers.address.zip;

        phs[0] := pers.phones[0];
        phs[1] := pers.phones[1];

        writeln;
        writeln('First name: ', pers.firstName);
        writeln(' Last name: ', pers.lastName);
        writeln('       Age: ', age);
        writeln('    Street: ', addr.street);
        writeln('      City: ', addr.city);
        writeln('     State: ', addr.state);
        writeln('       ZIP: ', addr.zip);
        writeln('  Phone #1: ', phs[0]);
        writeln('  Phone #2: ', phs[1]);
    END;

BEGIN
    john.firstName := 'John';
    john.lastName  := 'Doe';
    john.age := 24;
    john.address.street := '1680 25th Street';
    john.address.city   := 'San Pablo';
    john.address.state  := 'CALIFORNIA';
    john.address.zip    := '94806';
    john.phones[0]      := '111-1111';
    john.phones[1]      := '222-2222';

    age    := john.age;
    name   := john.firstName;
    street := john.address.street;

    writeln('   age = ', age);
    writeln('  name = ', name);
    writeln('street = ', street);

    persons[3].phones[1] := '888-8888';
    persons[2].phones[0] := persons[3].phones[1];
    writeln(' phone = ', persons[2].phones[0]);

    print(john);
END.

File: Mak_V3/Chapter17/src/RecordTest.java
import java.util.HashMap;
import java.util.Scanner;

public class RecordTest
{
    private static Scanner _standardIn = new Scanner(System.in);

    private static HashMap john;
    private static HashMap mary;

    private static void print(HashMap pers)
    {
        int age;
        HashMap addr = new HashMap();

        age = (Integer) pers.get("age");

        addr.put("street", (String) ((HashMap) pers.get("address")).get("street"));
        addr.put("city",   (String) ((HashMap) pers.get("address")).get("city"));
        addr.put("state",  (String) ((HashMap) pers.get("address")).get("state"));
        addr.put("zip",    (String) ((HashMap) pers.get("address")).get("zip"));

        System.out.println();
        System.out.println("First name: " + (String) pers.get("firstName"));
        System.out.println(" Last name: " + (String) pers.get("lastName"));
        System.out.println("       Age: " + age);
        System.out.println("    Street: " + (String) addr.get("street"));
        System.out.println("      City: " + (String) addr.get("city"));
        System.out.println("     State: " + (String) addr.get("state"));
        System.out.println("       ZIP: " + (String) addr.get("zip"));
    }

    public static void main(String[] args)
    {
        john = new HashMap();
        john.put("address", new HashMap());

        mary = new HashMap();
        mary.put("address", new HashMap());

        john.put("firstName", "John");
        john.put("lastName", "Doe");
        john.put("age", 24);
        ((HashMap) john.get("address")).put("street", "1680 25th Street");
        ((HashMap) john.get("address")).put("city", "San Pablo");
        ((HashMap) john.get("address")).put("state", "CA");
        ((HashMap) john.get("address")).put("zip", "94806");

        mary.put("firstName", "Mary");
        mary.put("lastName", "Jane");
        mary.put("age", 22);
        ((HashMap) mary.get("address")).put("street", "4899 Bela Drive");
        ((HashMap) mary.get("address")).put("city", "San Jose");
        ((HashMap) mary.get("address")).put("state", "CA");
        ((HashMap) mary.get("address")).put("zip", "95129");

        print(john);
        print(mary);
    }
}
